[HT-5] Added integration tests for backend

alexicon33 opened the pull request:
Задача: http://jira.prac.atp-fivt.org:8080/browse/HT-5
Добавил тесты и Readme с "документацией".

alexicon33 committed:
[HT-5] Added integration tests for backend

NikitaSikalov commented on server/app.ts:
```diff
@@ -33,20 +33,32 @@ dbHost = dbHost.replace(/{dbUser}/, dbUser);
+  console.log = () => {};
+  console.warn = () => {};
+  console.info = () => {};
+  console.error = () => {};
```
А это еще зачем?

NikitaSikalov replied:
Вместо этого у нас есть линтер, в котором ты можешь настроить правила.

alexicon33 replied:
Это для красоты: иначе между логами тестов выводятся всякие `deprecationWarning`-и самой `node.js` и левые `console.log`-и backend'a (оставлены они там не мной:) )
Про линтер не понял. (Я не хочу убирать console.log-и из кода, я просто не хочу, чтобы они выводились, когда `NODE_ENV === test`, а при разработке пусть выводятся)

NikitaSikalov replied:
Ну тогда перенаправляй выводы нативным `> /dev/null`  при запуске команды

alexicon33 replied:
Тогда не будет выводиться совсем ничего, а результаты тестов хотелось бы видеть.

NikitaSikalov commented on server/test/sample_data.ts:
```diff
@@ -0,0 +1,93 @@
+import { IProvider } from './../models/provider';
```
Это здорово что ты это написал, но Игнат как раз сделал задачу по добавлению фикстур данных в PR: https://github.com/HobbyTaste/web-application/pull/18. Тебе нужно синтегрировать эту систему в свой процесс тестирования

alexicon33 replied:
Я посмотрел, как у Игната сделаны фикстуры, но кажется, это не совсем то, что мне нужно. 
1) Библиотека `node-mongodb-fixtures`, которую он использовал, не очень дружит с typescript'ом — в ней в принципе нет файлов описания/объявления типов. 
2) Там, мягко говоря, довольно ограниченный набор действий: есть методы unload и load, удаляющие сразу все фикстуры из базы или загружающие их в неё с какой-то предобработкой. 
3) Они мне не очень нужны: мне не нужно ничего предварительно загружать в базу данных, наоборот, нужно, чтобы перед каждой группой тестов все коллекции были пустыми, а уж потом я буду сам посылать отдельные json'ы к бэкенду и смотреть, что там в базе появляется. Поэтому фикстуры я использовал только для очистки коллекций перед тестами.

NikitaSikalov replied:
1. Тебе и не нужно, чтобы эта библиотека дружила с `ts`. Тебе достаточно выполнять команду заливки фикстур перед каждым тестом в хуке `beforeEach`. Поможет запускать консольные команды из ts кода библиотека `shelljs`.
2. А каких методов тебе еще не хватает? Кажется Игнат должен был создать удобную инфраструктура для наливки нужных фикстур и данных в БД. Этим нужно пользоваться
3. Как я и писал в первом пункте, ты можешь вызывать наливку фикстур прям из кода тестов, так что не вижу с этим проблем.
Самое главное почему стоит пользоваться решением Игната, потому что мы будем поддерживать унифицированное решение одно.

NikitaSikalov commented on server/test/test.ts:
```diff
@@ -0,0 +1,457 @@
+    this.slow(10000);
+    before((done: Mocha.Done) => {
+        // Before group of tests we empty the database
+        Hobby.deleteMany({}, (err: Error) => {
```
Лучше писать не через колбэки, а через `async` и `await`. Так было бы более понятно.

NikitaSikalov commented on config/test.yaml:
```diff
@@ -0,0 +1,4 @@
+secrets:
```
Убираем, спроси у @SNVC1  про то как использовать БД из докер-контейнера для тестирования с предзаданными фикстурами

NikitaSikalov commented on .mocharc.json:
```diff
@@ -0,0 +1,5 @@
+{
+  "extension": ["ts"],
+  "spec": "server/test/test.ts",
```
У нас же будет не один файл с тестами

NikitaSikalov commented on package.json:
```diff
@@ -72,26 +72,33 @@
     "@typescript-eslint/eslint-plugin": "^2.6.0",
     "@typescript-eslint/parser": "^2.6.0",
     "antd": "^3.13.6",
+    "async": "^3.2.0",
```
Зачем?

alexicon33 replied:
В некоторых тестах мне надо было выполнить несколько асинхронных запросов к базе данных, а именно: запустить их, подождать, пока они все выполнятся, и только потом вызвать done(). В async ровно для этого есть метод parallel, который я и использовал.

NikitaSikalov replied:
А почему промисы не подошли?

NikitaSikalov commented on package.json:
```diff
@@ -13,7 +13,8 @@
     "lint": "./tasks/lint-staging",
     "lint:all": "yarn eslint server/ static/",
-    "mongo": "PATH=~/Mongo/mongodb-shell-linux-x86_64-ubuntu1804-4.2.2/mongodb-linux-x86_64-ubuntu1804-4.2.2/bin mongo \"mongodb+srv://hobbytaste-a3rpe.mongodb.net/test\"  --username sikalov"
+    "mongo": "PATH=~/Mongo/mongodb-shell-linux-x86_64-ubuntu1804-4.2.2/mongodb-linux-x86_64-ubuntu1804-4.2.2/bin mongo \"mongodb+srv://hobbytaste-a3rpe.mongodb.net/test\"  --username sikalov",
```
Это строчку не надо менять, ее вообще желательно было бы убрать

NikitaSikalov commented on package.json:
```diff
@@ -13,7 +13,8 @@
     "lint:all": "yarn eslint server/ static/",
-    "mongo": "PATH=~/Mongo/mongodb-shell-linux-x86_64-ubuntu1804-4.2.2/mongodb-linux-x86_64-ubuntu1804-4.2.2/bin mongo \"mongodb+srv://hobbytaste-a3rpe.mongodb.net/test\"  --username sikalov"
+    "mongo": "PATH=~/Mongo/mongodb-shell-linux-x86_64-ubuntu1804-4.2.2/mongodb-linux-x86_64-ubuntu1804-4.2.2/bin mongo \"mongodb+srv://hobbytaste-a3rpe.mongodb.net/test\"  --username sikalov",
+    "test": "NODE_ENV=test env TS_NODE_PROJECT='server/tsconfig.json' mocha --exit"
```
А `NODE_ENV` почему слева от env?

NikitaSikalov commented on server/app.ts:
```diff
@@ -33,20 +33,32 @@ dbHost = dbHost.replace(/{dbUser}/, dbUser);
 app.use(fileUpload());
 */
 
+// for testing
```
кажется комментарий этот не нужен

NikitaSikalov commented on server/app.ts:
```diff
@@ -82,6 +94,12 @@ try {
-        .then(() => logger.info(`Connect to mongoDB: success`))
+        .then(() => {
+          if (process.env.NODE_ENV !== 'test')
+           logger.info(`Connect to mongoDB: success`)
```
И что, так критично будет если при тестировании это сообщение вылезет?

alexicon33 replied:
Нет, просто мне это кажется некрасивым: если условие убрать, сначала выводится название первой группы тестов, потом это сообщение, а потом логи первого теста.

NikitaSikalov replied:
Лучше вообще убери эту строчку кода

NikitaSikalov commented on server/test/README.md:
```diff
@@ -0,0 +1,23 @@
+
+## Описание
+Все основные вещи, касающиеся тестов, сложены в директории `test`:
+- В файле `test.ts` описаны сами тесты;
```
Файлов должно предполагаться несколько, а не один

NikitaSikalov commented on server/test/README.md:
```diff
@@ -0,0 +1,23 @@
+## Описание
+Все основные вещи, касающиеся тестов, сложены в директории `test`:
+- В файле `test.ts` описаны сами тесты;
+- Файл `sample_data.ts` хранит маленький набор тестовых данных, использующихся для проверки работы контроллеров: 2 партнёра, 2 пользователя, 6 хобби и данные для обновления записей.
```
Опять же нужно связаться с @SNVC1

NikitaSikalov commented on server/test/test.ts:
```diff
@@ -0,0 +1,457 @@
+};
+
+
+describe('Sample Group', function() {
```
Тест просто огроменный - его точно нужно разбить на несколько describe и файлов

alexicon33 committed:
[HT-5] Fixed database, split tests

Теперь тесты работают с локальной базой данных, находящейся в `docker`-контейнере, они разделены на группы и исправлены так, чтобы они могли работать независимо.

NikitaSikalov commented on server/app.ts:
```diff
@@ -33,20 +33,31 @@ dbHost = dbHost.replace(/{dbUser}/, dbUser);
 app.use(fileUpload());
 */
 
+export const store = new MongoStore({
```
Перенеси это непосредственно в то место, где ты используешь это (то есть где импортируешь)

NikitaSikalov commented on server/test/README.md:
```diff
@@ -0,0 +1,23 @@
+
+## Описание
+Все основные вещи, касающиеся тестов, сложены в директории `test`:
+- Сами тесты описаны в файлах, начинающихся со слова "test". Сейчас их 3: `test_provider.ts`, `test_user.ts` и `test_hobby.ts`, в каждом из них описана соответствующая группа тестов.
```
Обычно делают вот такие файлы: `<name of test>.test.ts`

NikitaSikalov commented on server/test/test_hobby.ts:
```diff
@@ -0,0 +1,150 @@
+                useUnifiedTopology: true,
+                useNewUrlParser: true,
+            })
+            .then(() => utils.fixtures.unload())
```
лучше писать в нотации `async/await` - так более удобно и читабельно

NikitaSikalov commented on server/test/test_provider.ts:
```diff
@@ -0,0 +1,160 @@
+        let adding_functions: async.AsyncFunction<any>[] = [];
+        data.hobbies.forEach((hobby: Partial<IHobby>) => {
+            adding_functions.push(utils.create_hobby_adder(hobby));
+        });
```
`await Promise.all(data.hobbies.map(utils.create_hooby_adder)` - и всего одна строчка и async не нужен

alexicon33 committed:
[HT-5] Attached fixtures, used async/await

Переписал тесты через промисы и async/await, добавил использование фикстур, немного исправил их самих, добавил новые элементы.

NikitaSikalov committed:
add @types/shelljs
